lib/runtime-core/src/jit_debug.rs

```diff
@@ -0,0 +1,195 @@
+/// The data is in the form of a doubly linked list.  This global variable acts
+/// as a head node with extra information about the operation that we want the
+/// debugger to perform.
+#[no_mangle]
+#[allow(non_upper_case_globals)]
+pub(crate) static mut __jit_debug_descriptor: JitDebugDescriptor = JitDebugDescriptor {
+    version: 1,
+    action_flag: JITAction::JIT_NOACTION as _,
+    relevant_entry: ptr::null_mut(),
+    first_entry: ptr::null_mut(),
+};
+
+/// Prepend an item to the front of the `__jit_debug_descriptor` entry list
+///
+/// # Safety
+/// - Access to underlying global variable is unsynchronized.
+/// - Pointer to [`JITCodeEntry`] should point to a valid entry.
+unsafe fn push_front(jce: *mut JITCodeEntry) {
+    if __jit_debug_descriptor.first_entry.is_null() {
+        __jit_debug_descriptor.first_entry = jce;
+    } else {
+        let old_first = __jit_debug_descriptor.first_entry;
+        debug_assert!((*old_first).prev.is_null());
+        (*jce).next = old_first;
+        (*old_first).prev = jce;
+        __jit_debug_descriptor.first_entry = jce;
+    }
+}
+
+/// Removes an entry from the doubly linked list, updating both nodes that it's
+/// connected to.
+///
+/// # Safety
+/// - Access to underlying global variable is unsynchronized.
+/// - Pointer must point to a valid `JitCodeEntry`.
+unsafe fn remove_node(jce: *mut JITCodeEntry) {
+    if __jit_debug_descriptor.first_entry == jce {
+        debug_assert!((*jce).prev.is_null());
+        __jit_debug_descriptor.first_entry = (*jce).next;
+    }
+    if !(*jce).prev.is_null() {
+        (*(*jce).prev).next = (*jce).next;
+    }
+    if !(*jce).next.is_null() {
+        (*(*jce).next).prev = (*jce).prev;
+    }
+}
+
+/// Type for implementing Drop on the memory shared with the debugger.
+#[derive(Debug)]
+struct JITCodeDebugInfoEntryHandleInner(*mut JITCodeEntry);
+
+/// Handle to debug info about JIT code registered with a debugger
+#[derive(Debug, Clone)]
+pub(crate) struct JITCodeDebugInfoEntryHandle(Arc<JITCodeDebugInfoEntryHandleInner>);
+
+impl Drop for JITCodeDebugInfoEntryHandleInner {
+    fn drop(&mut self) {
+        unsafe {
+            // unregister the function when dropping the JIT code entry
+            __jit_debug_descriptor.relevant_entry = self.0;
+            __jit_debug_descriptor.action_flag = JITAction::JIT_UNREGISTER_FN as u32;
+            __jit_debug_register_code();
+            __jit_debug_descriptor.relevant_entry = ptr::null_mut();
+            __jit_debug_descriptor.action_flag = JITAction::JIT_NOACTION as u32;
+            remove_node(self.0);
+            let entry: Box<JITCodeEntry> = Box::from_raw(self.0);
+            Vec::from_raw_parts(
+                entry.symfile_addr,
+                entry.symfile_size as _,
+                entry.symfile_size as _,
+            );
+        }
+    }
+}
+
+/// Manager of debug info registered with the debugger.
+#[derive(Debug, Clone, Default)]
+pub struct JITCodeDebugInfoManager {
```

Comment:
```suggestion
pub(crate) struct JITCodeDebugInfoManager {
```

since there is only one field, which is private, and one method, which is visible to the crate only.

Reply:
It's part of module which has all public fields.  I wasn't sure if I should make the first private field in the struct or just make it public.  Making it public seemed like the safer choice